Generate Rust Jet modules

roconnor-blockstream opened the pull request.

roconnor-blockstream commented:
@apoelstra You don't need to review the code itself; however you should review the output.

apoelstra commented:
cc @chlewe 

I suggest we merge this as-is and then revisit it. It looks complete/correct as far as what it is, but it's incomplete -- in particular the `impl TryFrom<ElementsJetName> for &'static JetNode<Core>` block and similar also need to be autogenerated. Probably it'll take a bit of iteration to generate everything that we need to, and it might be better for you to do that based on this PR?

FYI with this PR you can run `cabal run GenRustJets` and it'll generate a bunch of `.rs` files in whatever directory you're in. Then you move those into `src/jet/init` in the Rust project and try to compile. Right now you get a ton of compilation errors because of references to the old names, but many of these look like code that should be autogenerated as well.

chlewe commented:
I'm also in favour of merging as is. In the future, we should reduce the number of required manual edits in Rust files.

I was thinking about merging the `*JetName` enum and `JetNode<*>` struct, but it is unclear how a Rust enum can contain the same data (CMR, types) for all enum variants. Maybe a `JetNodeInner` struct that is contained in all variants... The merge would make the `match` statements and `TryFrom` implementations clearer. The latter, we should automatically generate as part of the `init` files.

There was this idea of removing `TypeName`, but I do not see how we can represent types (as opposed to their names) in a const / static way. Rust is very restrictive here. We should add more names, like "byte", to shorten the generated names.

apoelstra reviewed (approved):
ACK 50ba60653a53d692574cdf58246558357dd9f0be

chlewe commented:
A range of jets seems to be missing from the generated files: `eq256`, `lt32`, `*_verify`, `sighash_all`, ...  Without these, the Miniscript compiler and basically all examples in rust-simplicity will not work. Are they missing because this is the first iteration, or has the way we use jets changed?

By the way, it would be cool if we had jets that handle bit strings of length `2^(2^n)` for arbitrary `n ≥ 0`, like in the jet catalogue. Not sure if this is possible.

roconnor-blockstream commented:
`SigAllHash` is there, but I'm missing a wrapper to build a standard signature scheme.

The other jets simply haven't been implemented as of yet.

roconnor-blockstream commented:
I've rebased this PR on some work in progress that adds jets for `CheckSig`, `Eq256`, `Le32`, and `Verify`.  Perhaps we will add  `EqVerify` and `LeVerify` jets down the road, but for now you will have to compose with the `Verify` jet.

As for the interpretation of Miniscript's checksig expression, as we should follow the [`checkSigVerify'` construction](https://github.com/ElementsProject/simplicity/blob/7b31f245fea74299fda00430ba08da30a751f160/Haskell/Core/Simplicity/Programs/CheckSig.hs#L59-L62) which is a combination of `disconnect` and a new `checkSigVerify` jet.

I've moved this PR into a draft state until the PRs this builds upto are completed and merged.  Until then, you can still use this branch for testing and experimentation.

roconnor-blockstream commented:
Okay, this is now rebased on master.  It is my understanding that more Rust code needs to be generated, so this isn't ready for merging yet.  However, there should now be more jets available.

uncomputable commented:
So, the `verify` jet is a shorthand for the `assert` construction? [What is `iden` doing in its definition?](https://github.com/ElementsProject/simplicity/blob/master/Haskell/Core/Simplicity/CoreJets.hs#L155) I would look this up myself, but I lack a Haskell IDE that allows me to jump to its definition _(suggestions appreciated)_.

uncomputable commented:
What is the value of `CheckSig` if all we do is `assert` / `verify` that its output is true? Is it a temporary jet until we have decided on how we bring `checkSigVerify'` into Rust?

uncomputable commented:
Speaking of `checkSigVerify'`, is it such a good idea to make use of `disconnect` in the standard conversion from Miniscript to Simplicity? Before universal sighashes make it into Bitcoin, this should be an optional add-on, I think. `disconnect` adds complexity that is currently not needed and it forces users to think about potential Bitcoin features that do not yet exist _(regarding how we teach this)_.

apoelstra commented:
@uncomputable I definitely would like universal sighashes in Elements. I think this is a super cool benefit of using Simplicity, which doesn't even require changing your upfront program, you just convert from Script to Simplicity and you're done.

Yes, there are open API/workflow questions about how to actually *use* universal sighashes but we can punt on that for now ... and if we instead implement some restricted form of sighash, then later we'll have to support two different `pk` equivalents in our simplicity-miniscript mapping.

roconnor-blockstream commented:
[`Verify`](https://github.com/ElementsProject/simplicity/blob/master/Haskell/Core/Simplicity/CoreJets.hs#L53) has type `Bit` to `()` and `Bit := () + ()`. The [`assert`macro](https://github.com/ElementsProject/simplicity/blob/2b29e59bccfc8f2682326306cc4b86ce51134268/Haskell/Core/Simplicity/Programs/Bit.hs#L50) process the right hand branch bringing the input type to `()`.  Then the `iden : () |- ()` expression passes that unit from the input to the output.  `unit : () |- ()` would also work just fine here.

---

In order to support batch verification of signatures, we only support `CheckSig` jets with verification semantics.  Tapscript does something similar.

---

I do think it is valuable to make `disconnect` part of the standard Miniscript translation.  Using `disconnect` means that UTXOs can immediately be used in covenants that require very fancy sighashes to operate. it is one of the best reasons to use Simplicity.  There are no real dangers of doing so.   The disconnected sighash programs must be signed, and so wallets shouldn't end up with any surprises.   I certainly think it is fine if wallets choose to only sign with a fix set of known programs.

uncomputable commented:
@roconnor-blockstream From what I understand, `Verify` is equivalent to `assert iden`. It is to be used as `t; Verify`. A more natural construction would be `assert t`, but this would be larger in program size. We cannot define `Verify` as `assert t`, because `t` cannot be jetted. This is why we resort to `assert iden`.

roconnor-blockstream commented:
One advantage of `assert t` is that it can leave it ambiguous as to whether there is a pruned branch there or not, but it comes at a substantial cost. `assert`s require a hidden node which is more than 32 bytes in size and maybe users aren't willing to pay that for branches that are pretty clearly actual asserts and not just a consequence of pruning.

roconnor-blockstream commented:
I've pushed a big update.  The encode and decode functions are still needed.  Probably other changes are needed too.  I'll revert to Draft for now, but feel free to leave comments.

sanket1729 commented on Haskell-Generate/GenRustJets.hs:
```diff
@@ -0,0 +1,346 @@
+
+rustWrapperImports :: Doc a
+rustWrapperImports = vsep (map (<> semi)
+  [ "use super::{frame_ffi::CFrameItem, jets_ffi, CTxEnv}"
```
For some reason, this is generating two semicolons. 
`use super::{frame_ffi::CFrameItem, jets_ffi};;`

roconnor-blockstream replied:
Done.

sanket1729 commented on Haskell-Generate/GenRustJets.hs:
```diff
@@ -0,0 +1,346 @@
+   , "}"
+   ]
+   where
+    envParam | CoreModule <- jetModule jet = "_env: const &CTxEnv"
```
Here and below: There should not be any const

roconnor-blockstream replied:
Done.

sanket1729 commented on Haskell-Generate/GenRustJets.hs:
```diff
@@ -0,0 +1,346 @@
+  wrapper (SomeArrow jet) = vsep
+   [ nest 4 $ vsep
+     [ pretty $ "pub fn "++cJetName jet++"(dst: &mut CFrameItem, src: CFrameItem, "++envParam++") -> bool {"
+     , pretty $ "unsafe { "++lowerRustModuleName mod++"_ffi::"++cJetName jet++"(dst, src, "++envArg++") }"
```
should be `jets_ffi::` instead of `elements_ffi::`

sanket1729 replied:
Given the application specific nature of core jets, it makes sense to have `elements_ffi` instead of `jets_ffi`. Will update rust code to reflect that

sanket1729 commented on Haskell-Generate/GenRustJets.hs:
```diff
@@ -0,0 +1,346 @@
+  , "use crate::merkle::cmr::Cmr"
+  , "use crate::{decode_bits, Error}"
+  , "use bitcoin_hashes::sha256::Midstate"
+  , "use simplicity_sys::c_jets::CTxEnv"
```
This line is no longer needed

roconnor-blockstream replied:
Done.

sanket1729 commented on Haskell-Generate/GenRustJets.hs:
```diff
@@ -0,0 +1,346 @@
+  modname = rustModuleName mod
+
+rustJetTy fname getTy mod = vsep $
+  [ nest 4 (vsep (pretty ("fn "++fname++"(&self) -> TypeName {") :
```
These should not  have `TypeName` around it

```diff
- Core::Add32 => TypeName(b"*2i"),
+ Core::Add32 => b"*2i"
```

roconnor-blockstream replied:
done.

sanket1729 commented on Haskell-Generate/GenRustJets.hs:
```diff
@@ -0,0 +1,346 @@
+    [ nest 4 (vsep ("match self {" :
+        map (<>comma)
+        [ pretty modname <> "::" <> pretty (jetName jet) <+> "=>" <+>
+          pretty ("&simplicity_sys::c_jets::jet_wrapper::"++cJetName jet)
```
jet_wrapper should jets_wrapper

roconnor-blockstream replied:
done.

sanket1729 reviewed (commented):
Could not check elements.rs file because it generates all the jets. Left comments about `jets_wrapper.rs`, `core.rs`. `jets_ffi.rs` seems to be fine

sanket1729 commented:
Files core.rs need import
```
use simplicity_sys::CFrameItem;
```
File elements.rs needs import 
```
use simplicity_sys::{CFrameItem, CElementsTxEnv};
use crate::jet::elements::ElementsEnv;
```

sanket1729 commented:
Apart from the above changes, everything is good.

sanket1729 commented on Haskell-Generate/GenRustJets.hs:
```diff
@@ -271,16 +271,23 @@ rustHeader = "/* This file has been automatically generated. */"
   , "use crate::merkle::cmr::Cmr"
   , "use crate::{decode_bits, Error}"
   , "use bitcoin_hashes::sha256::Midstate"
+  , "use simpicity_sys::CFrameItem"
```
/s/simpicity/simplicity

roconnor-blockstream replied:
Done.

sanket1729 reviewed (commented):
One last spelling error. Everything else if good.

sanket1729 commented on Haskell-Generate/GenRustJets.hs:
```diff
@@ -0,0 +1,377 @@
+rustWrapperImports :: Doc a
+rustWrapperImports = vsep (map (<> semi)
+  [ "use crate::CElementsTxEnv"
+  , "use super::{frame_ffi::CFrameItem, jets_ffi}"
```
Promise this is the last functional thing. 
```diff
- "use super::{frame_ffi::CFrameItem, jets_ffi}"
+ "use super::{frame_ffi::CFrameItem, elements_ffi}"
```

roconnor-blockstream replied:
Done.

sanket1729 reviewed (commented):
This is the last one for sure

sanket1729 commented:
ACK, the rust generated code is compiles and builds successfully.

sanket1729 reviewed (commented):
Sorry about the review again. Requesting some changes would allow us to have our encoding/decoding temporarily run using the manual implementation we already have. 

This is useful in testing the simplicity program we create in rust via the FFI interface.

roconnor-blockstream commented on Haskell-Generate/GenRustJets.hs:
```diff
@@ -0,0 +1,377 @@
+  , "use crate::jet::type_name::TypeName"
+  , "use crate::jet::Jet"
+  , "use crate::merkle::cmr::Cmr"
+  , "use crate::{decode_bits, Error}"
```
Done.

roconnor-blockstream commented on Haskell-Generate/GenRustJets.hs:
```diff
@@ -0,0 +1,377 @@
+
+  encodeStub = vsep
+    [ "fn encode<W: Write>(&self, w: &mut BitWriter<W>) -> std::io::Result<usize> {"
+    , "    todo!()"
```
Done.

uncomputable commented:
Two nits about the currently generated code:
1. `cargo fmt` changes the code. Either we fix this or we run `cargo fmt` after each generation.
  - import order should be alphanumeric
  - long strings (`target_ty`) or function calls (`c_jet_ptr`) should be wrapped
2. The `Jet` methods are not in the same [order as in the trait](https://github.com/ElementsProject/rust-simplicity/blob/master/src/jet/mod.rs#L84-L107). `cargo clippy` doesn't complain about this but it is shown in my IDE.

uncomputable commented:
`cargo clippy` also complains about unused imports and unused function arguments. @roconnor-blockstream I could push a fixup commit (somehow) if this doesn't destroy your workflow.

sanket1729 commented:
@uncomputable, I think we should figure out a way to disable `rustfmt` formatting lints for auto-generated code. Are you testing the latest code? I don't see any unused args or unused imports warnings

uncomputable commented:
I was using an older version. The clippy warnings seem to be gone. As for rustfmt, the import and method order are easy to fix. The wrapped return values are annoying, because shorter values _must not_ be wrapped, so I agree that we should disable the formatter.

sanket1729 commented on Haskell-Generate/GenRustJets.hs:
```diff
@@ -0,0 +1,448 @@
+  , "use crate::decode_bits"
+  , "use bitcoin_hashes::sha256::Midstate"
+  , "use simplicity_sys::CFrameItem"
+  , "use std::io::Write"
```
Needs "use crate::Error"

roconnor-blockstream replied:
Done.

sanket1729 reviewed (commented):
Just one missing piece in enc/decoding APIs

sanket1729 commented:
The output rust code is compiling correctly. Are there any other features lacking for this PR?

roconnor-blockstream commented:
The Haskell pseudo-PR at the base of this PR needs tidying up.  I actually forgot about this, so I'll get on it.

roconnor-blockstream commented:
Rebased and squshed.

roconnor-blockstream commented:
Generated output is identical to 1d03dc9624ff242802d28bfdf7fb5562d40b71d4.
